ShaderSwitch Improvements

johnhaddon opened the pull request:
This makes big improvements to the ShaderSwitch so that it is much more useful with Arnold and OSL shaders. It can now switch the input for parameters of any type whereas before it could only switch entire shaders and RSL coshaders. It can now also have a context-varying index value, so expressions can be used to drive the index for the first time.

The low level work here was actually pretty straightforward, since I'd planned for it when rejigging the NetworkGenerator to support pass-throughs for OSL shaders. The UI needed much more work though. Whereas before the ShaderSwitch could just add the "in" and "out" plugs during construction, we must now wait until the user first wants to make a connection, because only then do we know what sort of plug is needed. I've dealt with this by introducing a new PlugAdder gadget, that appears as a + icon on the side of the node. Dragging to or from this triggers the addition of the relevant plugs and the creation of a connection, as show below...

![switch](https://cloud.githubusercontent.com/assets/1133871/20794399/0772a4c8-b7c4-11e6-836e-dccd62e909cf.png)

The implementation of this is a little nasty because of the required interaction between PlugAdder, StandardNodule and StandardNodeGadget, but I don't see any other way without a significant redesign, and the drag snapping that required the interactions is essential for a decent experience. I'll be capitalising on this bit of ickiness as much as possible in the future by extending the PlugAdder to add plugs to Boxes and Dots, and for connecting to hidden plugs in the upcoming please-don't-show-me-all-of-alsurface-at-once feature. I expect the PlugAdder implementation to change a bit to allow these customisations (maybe becoming an abstract base class), but I didn't want to second guess what the right design was until I started working on the additional use cases.

Finally, all this work means that we can now add a generic Switch node to the Utility menu, which can be set up to work with any plug type. In fact, you could argue that the specialised switch types are no longer actually necessary. User polling didn't really reveal a preference one way or the other, so the specialised switches remain for now - I suspect we need to get the generic Switch into people's hands until they'll know what they prefer.

johnhaddon commented:
Would be good to get this one reviewed today if possible, since my next PR will depend on it.

andrewkaufman commented on src/GafferUI/PlugAdder.cpp:
```diff
@@ -0,0 +1,354 @@
+
+		updateMetadata( inPlug, "nodeGadget:nodulePosition", edgeName( inEdge ) );
+		updateMetadata( inPlug, "compoundNodule:orientation", edgeOrientation( inEdge ) );
+		updateMetadata( outPlug, "nodeGadget:nodulePosition", edgeName( oppositeEdge( inEdge ) ) );
```
Should we be updating "nodule:color" and/or "connectionGadget:color" here too? At IE I'm seeing some funky color behaviour with Color3fPlugs and FloatPlugs (though TaskPlug, ImagePlug, ScenePlug, and FilterPlug seem to be doing what I'd expect). I'd also be handy for Jabuka EntityPlugs (which are really just StringPlugs but have instance metadata saying the plug and its connections are green).

andrewkaufman replied:
So at IE we do specify color (both nodule and connection) by plug type: https://github.com/ImageEngine/gafferExtensions/blob/master/config/lightingTools/colorScheme.py#L20-L45

And thats resulting in "what I'd expect" for scene/image/filter/task:

![switchcolors](https://cloud.githubusercontent.com/assets/3979682/20906897/4dcd6046-baff-11e6-9774-1bae1a5de839.png)

But something "funky" for color and float (despite being specified in that same section of the colorScheme config). I guess those colors are being overridden by this https://github.com/johnhaddon/gaffer/blob/12b2dc4f495779ac9cf0710dd1d50965a6ce1ca8/startup/gui/nodeGraph.py#L84-L88, but wouldn't that apply to the original shader as well?

![switchcolorsfunky](https://cloud.githubusercontent.com/assets/3979682/20906946/b1373f12-baff-11e6-8efb-1d27b490a33e.png)

andrewkaufman replied:
Wait, maybe I've got that backwards and the "funky" colors are the ones coming from the IE config, which were rarely seen before, since color and float nodules are exposed almost exclusively by shaders... so maybe we should just update our coding to use the same colors that gaffer is setting for shader plugs?

It still would be nice to have the instance colors from the Jabuka nodes transfer over though...

andrewkaufman commented on src/GafferUI/PlugAdder.cpp:
```diff
@@ -0,0 +1,354 @@
+
+bool PlugAdder::buttonPress( GadgetPtr gadget, const ButtonEvent &event )
+{
+	return event.buttons == ButtonEvent::Left;
```
I'm having some issues with right mouse button press (on the + gadget and on the node itself):

```
Traceback (most recent call last):
  File "/home/andrewk/apps/gaffer/0.30.2.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/Widget.py", line 903, in eventFilter
    return self.__mouseButtonPress( qObject, qEvent )
  File "/home/andrewk/apps/gaffer/0.30.2.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/Widget.py", line 1003, in __mouseButtonPress
    result = widget._buttonPressSignal( widget, event )
RuntimeError: Exception : Traceback (most recent call last):
  File "/home/andrewk/apps/gaffer/0.30.2.0dev/cent7.x86_64/cortex/9/gaffer/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/home/andrewk/apps/gaffer/0.30.2.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/NodeGraph.py", line 268, in __buttonPress
    self.nodeContextMenuSignal()( self, nodeGadget.node(), overrideMenuDefinition )
AttributeError: 'GraphComponent' object has no attribute 'node'
```

andrewkaufman replied:
Also, it might be nice if left mouse button press allowed me to setup the switch as a scene/image/filter/color/whatever. Since its a + button its fairly intuitive to want to click it directly.

johnhaddon replied:
I've reproduced the right click issue here, and I think I see what's going on. Looks like I might have to make SwitchNodeGadget public and bind it, which would be a pity. I'll see if I have any better ideas overnight.

The left click idea sounds good to me - could we defer that till I have my next Shader PR out of the way though? I already have changes on top of PlugAdder for that PR, and it'll be easier if I can avoid adding intermediate changes in.

andrewkaufman replied:
Sure

andrewkaufman commented on python/GafferUI/SwitchUI.py:
```diff
@@ -58,9 +58,23 @@
+				by the index plug to be passed through to the output.
+				""",
+
+				"nodule:type", "GafferUI::CompoundNodule",
```
Don't most input arrays get some extra spacing argument? The nodules on `Switch` seem really close together compared to a `SceneSwitch` or `ImageSwitch`. Actually, now that I look at it, those ones are really wide compared to `TaskSwitch` and `TaskList`...

andrewkaufman replied:
Also the "in" array shouldn't have a widget in the NodeEditor. "in.*" and "out" should only have one if the plug they were setup from does.

andrewkaufman replied:
I was going to suggest that originally, but then I wasn't sure if having the widget for an out color was useful?

johnhaddon commented:
OK, I've updated as follows :

- Hid "in" and "out" widgets in the NodeEditor (ca17ec6)
- Fixed right click NodeEditor bug (390e740)

I think you're right that those funky colours are coming from the IE config - I deliberately avoided any bright blues in the Gaffer config because they don't play well with the highlight colour.

If it's OK with you, I'd like to defer the colour/spacing questions till I've got the next PR out of the way. Any monkeying with the metadata that would necessitate will be in the code I've already rejigged to support custom PlugAdders.

andrewkaufman commented:
Ok, so should I remake those color/spacing comments on the next PR?

johnhaddon commented:
> Ok, so should I remake those color/spacing comments on the next PR?

This next PR is getting pretty involved itself (big refactor of StandardNodeGadget and CompoundNodule to share the same layout code). I think what would be good would be a followup where we do the colours/spacing and also use the PlugAdder to improve the Dot and Box UI, but if you feel it's urgent then I can squeeze the first two in...

andrewkaufman commented:
I don't feel like its urgent per se, just that we shouldn't have a release without it really, and I'm doubting my ability to keep track of 3-4 random requests I made on a PR that's already merged.

johnhaddon commented:
I've added it on to my todo list to keep a track of it.
